edge-util/src/Microsoft.Azure.Devices.Edge.Util/metrics/prometheus.net/MetricsProvider.cs

```diff
@@ -8,22 +8,40 @@ namespace Microsoft.Azure.Devices.Edge.Util.Metrics.Prometheus.Net
     using System.Threading;
     using System.Threading.Tasks;
     using global::Prometheus;
+    using Microsoft.Extensions.Logging;
 
     public class MetricsProvider : IMetricsProvider
     {
         const string CounterNameFormat = "{0}_{1}_total";
         const string NameFormat = "{0}_{1}";
-        const string InstanceFile = "metrics_instance";
+        const string InstanceFileName = "instance";
+        const string InstanceFolderName = "metrics_instance";
+        readonly string instanceFileAbsolutePath;
         readonly string namePrefix;
         readonly List<string> defaultLabelNames;
-        static readonly Lazy<string> instanceNumber = new Lazy<string>(GetInstanceNumber);
+        readonly ILogger logger = Logger.Factory.CreateLogger<MetricsProvider>();
 
-        public MetricsProvider(string namePrefix, string iotHubName, string deviceId)
+        public MetricsProvider(string namePrefix, string iotHubName, string deviceId, string storagePath)
         {
             this.namePrefix = Preconditions.CheckNonWhiteSpace(namePrefix, nameof(namePrefix));
+
             Preconditions.CheckNonWhiteSpace(iotHubName, nameof(iotHubName));
             Preconditions.CheckNonWhiteSpace(deviceId, nameof(deviceId));
-            this.defaultLabelNames = new List<string> { iotHubName, deviceId, instanceNumber.Value };
+            Preconditions.CheckNonWhiteSpace(storagePath, nameof(storagePath));
+
+            string storageDirectory = Path.Combine(storagePath, InstanceFolderName);
+            try
+            {
+                Directory.CreateDirectory(storageDirectory);
+                this.instanceFileAbsolutePath = Path.Combine(storageDirectory, InstanceFileName);
+            }
+            catch (Exception ex)
+            {
+                this.logger.LogError(ex, "Could not create metrics directory");
+            }
+
+            string instanceNumber = this.GetInstanceNumber();
```

Comment:
What happens if GetInstanceNumber returns zero?

Reply:
My understanding is that it will continue the normal flow thereby signalling a failed read of the instance file.